src/cocoa/toga_cocoa/widgets/table.py

```diff
@@ -21,58 +25,61 @@ def numberOfRowsInTableView_(self, table) -> int:
         return len(self.interface.data) if self.interface.data else 0
 
     @objc_method
-    def tableView_objectValueForTableColumn_row_(self, table, column, row: int):
+    def tableView_viewForTableColumn_row_(self, table, column, row: int):
+
         data_row = self.interface.data[row]
+        col_identifier = str(column.identifier)
 
         try:
-            # Obtain the _impl for the data row...
-            data = data_row._impl
+            value = getattr(data_row, col_identifier)
+
+            # if the value is a widget itself, just draw the widget!
+            if isinstance(value, toga.Widget):
+                return value._impl.native
+
+            # Allow for an (icon, value) tuple as the simple case
+            # for encoding an icon in a table cell. Otherwise, look
+            # for an icon attribute.
+            elif isinstance(value, tuple):
+                icon_iface, value = value
+            else:
+                try:
+                    icon_iface = value.icon
+                except AttributeError:
+                    icon_iface = None
         except AttributeError:
-            # or if it doesn't already exist, create it.
-            data = {}
-            data_row._impl = data
+            # If the node doesn't have a property with the
+            # accessor name, assume an empty string value.
+            value = self.interface.missing_value
```

Comment:
This can throw a ValueError (and will by default); that needs to be caught and output to console as a warning (line 53-56 of the original)